File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
  bool matches(int map1[],int map2[]){
    for(int i = 0;i < 26;i ++){
      if(map1[i] != map2[i]){
        return false;
      }
    }
    return true;
  }
 public:
  bool checkInclusion(string s1, string s2) {
    if (s1.length() > s2.length()) return false;
    int map1[26] = {0};
    for (int i = 0; i < s1.length(); i++) {
      map1[s1[i] - 'a']++;
    }
    for (int i = 0; i <= s2.length() - s1.length(); i++) {
      int map2[26] = {0};
      for (int j = 0; j <  s1.length(); j++) {
        map2[s2[i+j] - 'a']++;
      }
      if(matches(map1,map2)){
        return true;
      }
    }
    return false;
  }
};

File: 0567-permutation-in-string/0567-permutation-in-string.py
# time complexity: O(n)
# space complexity: O(n)
from typing import Counter


class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        n1 = len(s1)
        n2 = len(s2)
        counter1 = Counter(s1)
        counter2 = Counter(s2[:n1])
        if counter1 == counter2:
            return True
        for i in range(n2 - n1 ):
            leftChar = s2[i]
            rightChar = s2[i + n1]
            counter2[leftChar] -= 1
            counter2[rightChar] += 1
            if counter1 == counter2:
                return True

        return False


s1 = "ab"
s2 = "eidbaooo"
print(Solution().checkInclusion(s1, s2))
s1 = "ab"
s2 = "eidboaoo"
print(Solution().checkInclusion(s1, s2))
s1 = "adc"
s2 = "dcda"
print(Solution().checkInclusion(s1, s2))
